File: leetcode/109/src/P109.java
/**
 * Created by qwordy on 2/24/16.
 * 109. Convert Sorted List to Binary Search Tree
 */

class ListNode {
	int val;
	ListNode next;
	ListNode(int x) { val = x; }
}

class TreeNode {
	int val;
	TreeNode left;
	TreeNode right;
	TreeNode(int x) { val = x; }
}

class Solution {
	public TreeNode sortedListToBST(ListNode head) {
		int len = 0;
		ListNode p = head;
		while (p != null) {
			p = p.next;
			len++;
		}
		node = head;
		return build(0, len - 1);
	}

	private TreeNode build(int l, int r) {
		if (l > r) return null;
		int mid = (l + r) / 2;
		TreeNode left = build(l, mid - 1);
		TreeNode root = new TreeNode(node.val);
		node = node.next;
		root.left = left;
		root.right = build(mid + 1, r);
		return root;
	}

	private ListNode node;
}

public class P109 {
}

File: leetcode/109/P109.rb
# 109. Convert Sorted List to Binary Search Tree
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {ListNode} head
# @return {TreeNode}
def sorted_list_to_bst(head)
	len = 0
	p = head
	while p
		p = p.next
		len += 1
	end
	$node = head
	build 0, len - 1
end

def build l, r
	return nil if l > r
	mid = (l + r) / 2
	left = build l, mid - 1
	root = TreeNode.new $node.val
	$node = $node.next
	root.left = left
	root.right = build mid + 1, r
	root
end

